nodes/arm_node.cpp

```diff
@@ -1,110 +1,247 @@
+  void updateCartesianWaypoints(hebi_cpp_api_examples::TargetWaypoints target_waypoints) {
+    action_server_->setAborted();
+
+    // Fill in an Eigen::Matrix3xd with the xyz goal
+    size_t num_waypoints = target_waypoints.waypoints_vector.size();
+    Eigen::Matrix3Xd xyz_waypoints(3, num_waypoints);
+    for (size_t i = 0; i < num_waypoints; ++i) {
+      const auto& xyz_waypoint = target_waypoints.waypoints_vector[i];
+      xyz_waypoints(0, i) = xyz_waypoint.x;
+      xyz_waypoints(1, i) = xyz_waypoint.y;
+      xyz_waypoints(2, i) = xyz_waypoint.z;
+    }
+
+    // Replan
+    Eigen::Matrix3Xd tip_directions(3, 0);
+    updateCartesianWaypoints(xyz_waypoints, tip_directions);
+  }
 
-        // These are the joint angles that will be added
-        Eigen::MatrixXd positions(arm_.size(), num_waypoints);
+  // Set the target end effector location in (x,y,z) space, replanning
+  // smoothly to the new location
+  void setTargetCallback(geometry_msgs::Point data) {
+    action_server_->setAborted();
 
-        // Plan to each subsequent point from the last position
-        Eigen::VectorXd last_position = arm_.getLastFeedback().getPosition();
+    // Fill in an Eigen::Matrix3xd with the xyz goal
+    Eigen::Matrix3Xd xyz_waypoints(3, 1);
+    xyz_waypoints(0, 0) = data.x;
+    xyz_waypoints(1, 0) = data.y;
+    xyz_waypoints(2, 0) = data.z;
 
-        // Get joint angles to move to each waypoint
-        for (size_t i = 0; i < num_waypoints; ++i) {
-          const auto& waypoint = target_waypoints.waypoints_vector[i];
+    // Replan
+    Eigen::Matrix3Xd tip_directions(3, 0);
+    updateCartesianWaypoints(xyz_waypoints, tip_directions);
+  }
 
-          Eigen::Vector3d xyz(waypoint.x, waypoint.y, waypoint.z);
+  // "Jog" the target end effector location in (x,y,z) space, replanning
+  // smoothly to the new location
+  void offsetTargetCallback(geometry_msgs::Point data) {
+    action_server_->setAborted();
+
+    // Only update if target changes!
+    if (data.x == 0 && data.y == 0 && data.z == 0)
+      return;
+
+    // Initialize target from feedback as necessary
+    if (!isTargetInitialized()) {
+      auto pos = arm_.getKinematics().FK(arm_.getLastFeedback().getPositionCommand());
+      target_xyz_.x() = pos.x();
+      target_xyz_.y() = pos.y();
+      target_xyz_.z() = pos.z();
+    }
 
-          // Find the joint angles for the next waypoint, starting from the last
-          // waypoint
-          last_position = arm_.getKinematics().solveIK(last_position, xyz);
+    // Fill in an Eigen::Matrix3xd with the xyz goal
+    Eigen::Matrix3Xd xyz_waypoints(3, 1);
+    xyz_waypoints(0, 0) = target_xyz_.x() + data.x;
+    xyz_waypoints(1, 0) = target_xyz_.y() + data.y;
+    xyz_waypoints(2, 0) = target_xyz_.z() + data.z;
 
-          // Save the waypoints
-          positions.col(i) = last_position; 
-        }
+    // Replan
+    Eigen::Matrix3Xd tip_directions(3, 0);
```

Comment:
This seems like a code smell. Why do you have to pass in an empty matrix?

Reply:
fixed